src/v/pandaproxy/schema_registry/protobuf.cc

```diff
@@ -171,12 +174,34 @@ ss::future<const pb::FileDescriptor*> import_schema(
   sharded_store& store,
   const canonical_schema& schema) {
     try {
-        co_return co_await build_file_with_deps(dp, store, schema);
+        co_return co_await build_file_with_refs(dp, store, schema);
     } catch (const exception& e) {
         throw as_exception(invalid_schema(schema));
     }
 }
 
+///\brief Build a FileDescriptor and import references from the store.
+///
+/// Recursively import references into the DescriptorPool, building the files
+/// on stack unwind.
+ss::future<const pb::FileDescriptor*> build_file_with_refs(
```

Comment:
Do we need to worry about cycles here?  Can two schemas reference each other and hang redpanda that way?

Reply:
> Do we need to worry about cycles here? Can two schemas reference each other and hang redpanda that way?

Good question. I see two scenarios:
1) The first subject is added to the store with a reference to the second. Building fails as second doesn't exist yet.
2) The first subject is added, the second subject is added with a reference to the first, the first is updated with a reference to the second. This update fails to build as the first subject already exists (`DescriptorPool` doesn't like duplicates). I could [add  a check](https://github.com/vectorizedio/redpanda/pull/2822/files#diff-ab5f89e82c8c100dee5ac7eda1caeb003b1f2055da08789a9a21d952e1f16e8fR219), as I do in `build_file_with_deps`, if you think that makes it clearer/safer?

In a real schema registry, adding the schema would fail before it hits the store because it's validated first, which calls build.

It does bring up a good question about versioning support, but Protobuf doesn't support versions, so name collisions occur. I think this is OK. A `protobuf_schema_definition` has it's own isolated `DescriptorPool` to prevent unnecessary collisions.

Reply:
OK, and in case 2 it's also impossible for someone to delete & recreate the first subject with a reference to second, right?  (i.e. we prevent deleting referred-to subjects)

Reply:
> OK, and in case 2 it's also impossible for someone to delete & recreate the first subject with a reference to second, right? (i.e. we prevent deleting referred-to subjects)

Oh, that might allow breaking things. References aren't currently fully supported, and deleting a referenced subject isn't prevented. I'll check the required behaviour here.

Reply:
I'll add this in another PR, Protobuf support isn't advertised yet (`GET /schemas/types/`)